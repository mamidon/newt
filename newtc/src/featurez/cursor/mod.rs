
mod cursor;
mod tests;

pub use self::cursor::Cursor;
