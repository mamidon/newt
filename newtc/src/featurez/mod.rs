mod cursor;
mod driver;
mod grammar;
mod newtypes;
mod parse;
mod runtime;
mod syntax;
mod tokens;

pub use self::runtime::VirtualMachine;
pub use self::syntax::SyntaxTree;
pub use self::tokens::{tokenize, StrTokenSource, Token, TokenKind};
