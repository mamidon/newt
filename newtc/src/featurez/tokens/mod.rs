mod tests;
mod token;
mod token_source;
mod tokenize;
mod tokenkind;

pub use self::token::*;
pub use self::token_source::StrTokenSource;
pub use self::tokenize::tokenize;
pub use self::tokenkind::*;
