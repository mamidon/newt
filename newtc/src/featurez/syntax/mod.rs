use super::tokens::{Token, TokenKind};
use super::newtypes::TransparentNewType;

use std::rc::Rc;
use std::fmt::Display;
use std::fmt::Formatter;
use std::fmt::Error;

mod syntax_kind;
mod syntax_node;
mod syntax_token;
mod syntax_element;
mod syntax_tree;
mod token_source;
mod tree_sink;
mod text_tree_sink;
mod tests;

pub use self::syntax_kind::SyntaxKind;
pub use self::syntax_node::SyntaxNode;
pub use self::syntax_token::SyntaxToken;
pub use self::syntax_element::SyntaxElement;
pub use self::syntax_tree::SyntaxTree;
pub use self::token_source::TokenSource;
pub use self::tree_sink::TreeSink;
pub use self::text_tree_sink::TextTreeSink;


#[repr(transparent)]
pub struct LiteralExprNode(SyntaxNode);

unsafe impl TransparentNewType for LiteralExprNode {
	type Inner = SyntaxNode;
}

#[repr(transparent)]
pub struct BinaryExprNode(SyntaxNode);

impl BinaryExprNode {
	pub fn lhs(&self) -> &BinaryExprNode {
		let lhs_node = self.0.nth_node_kind(0, SyntaxKind::LiteralExpr);
		BinaryExprNode::from_inner(lhs_node)
	}
}

unsafe impl TransparentNewType for BinaryExprNode {
	type Inner = SyntaxNode;
}
