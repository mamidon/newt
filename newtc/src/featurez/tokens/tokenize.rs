use crate::featurez::cursor::{Cursor};
use crate::featurez::syntax::TokenSource;
use crate::featurez::tokens::{
	TokenKind,
	Token,
};

pub fn tokenize(text: &str) -> Vec<Token> {
	let mut tokens: Vec<Token> = vec![];
	let mut source = text;

	while !source.is_empty() {
		let mut token = next_token(source);

		tokens.push(token);
		source = &source[token.lexeme_length()..];
	}

	tokens.push(Token::new(TokenKind::EndOfFile, 0));
	tokens
}

fn next_token(text: &str) -> Token {
	let mut cursor = Cursor::new(text);

	if let Some(token) = lex_whitespace(&mut cursor) {
		token
	} else if let Some(token) = lex_multi_character_token(&mut cursor) {
		token
	} else if let Some(token) = lex_two_character_token(&mut cursor) {
		token
	} else if let Some(token) = lex_single_character_token(&mut cursor) {
		token
	} else {
		cursor;
		Token::new(TokenKind::TombStone, 1)
	}
}


fn lex_whitespace(cursor: &mut Cursor) -> Option<Token> {
	while cursor.current().is_some() && cursor.match_char_predicate(&|c: char| c.is_whitespace()) {
		cursor.next();
	}

	if cursor.len() > 0 {
		Some(Token::new(TokenKind::WhiteSpace, cursor.len()))
	} else {
		None
	}
}


fn lex_single_character_token(cursor: &mut Cursor) -> Option<Token> {
	fn make_token(token_kind: TokenKind) -> Token {
		Token::new(token_kind, 1)
	}

	let current = cursor.current();

	let token = match current {
		Some('{') => make_token(TokenKind::LeftBrace),
		Some('}') => make_token(TokenKind::RightBrace),
		Some('(') => make_token(TokenKind::LeftParenthesis),
		Some(')') => make_token(TokenKind::RightParenthesis),
		Some('[') => make_token(TokenKind::LeftBracket),
		Some(']') => make_token(TokenKind::RightBracket),

		Some(',') => make_token(TokenKind::Comma),
		Some('.') => make_token(TokenKind::Dot),
		Some(':') => make_token(TokenKind::Colon),
		Some(';') => make_token(TokenKind::SemiColon),
		Some('_') => make_token(TokenKind::UnderScore),

		Some('=') => make_token(TokenKind::Equals),
		Some('+') => make_token(TokenKind::Plus),
		Some('-') => make_token(TokenKind::Minus),
		Some('*') => make_token(TokenKind::Star),
		Some('/') => make_token(TokenKind::Slash),
		Some('>') => make_token(TokenKind::Greater),
		Some('<') => make_token(TokenKind::Less),

		Some('|') => make_token(TokenKind::Pipe),
		Some('&') => make_token(TokenKind::Ampersand),
		Some('!') => make_token(TokenKind::Bang),

		_ => return None
	};

	cursor.next();

	Some(token)
}


fn lex_two_character_token(cursor: &mut Cursor) -> Option<Token> {
	fn make_token(cursor: &mut Cursor, token_kind: TokenKind) -> Token {
		cursor.next();
		cursor.next();

		Token::new(token_kind, 2)
	}

	fn make_token_consume_line(cursor: &mut Cursor, token_kind: TokenKind) -> Token {
		let mut length = 2;
		cursor.next();
		cursor.next();

		while cursor.current().is_some() && cursor.match_char_predicate(&|c: char| c != '\n') {
			length += 1;
			cursor.next();
		}

		Token::new(token_kind, length)
	}

	let current = cursor.current();
	let next = cursor.peek(1);

	if let (Some(current), Some(next)) = (cursor.current(), cursor.peek(1)) {
		let token = match (current, next) {
			('=', '=') => make_token(cursor, TokenKind::EqualsEquals),
			('>', '=') => make_token(cursor, TokenKind::GreaterEquals),
			('<', '=') => make_token(cursor, TokenKind::LessEquals),
			('|', '|') => make_token(cursor, TokenKind::PipePipe),
			('&', '&') => make_token(cursor, TokenKind::AmpersandAmpersand),
			('/', '/') => make_token_consume_line(cursor, TokenKind::CommentLine),
			_ => return None
		};

		Some(token)
	} else {
		None
	}
}



fn lex_multi_character_token(cursor: &mut Cursor) -> Option<Token> {
	if let Some(numeric_literal) = scan_numeric_literal(cursor) {
		Some(numeric_literal)
	} else if let Some(string) = scan_string_literal(cursor) {
		Some(string)
	} else if let Some(glyph) = scan_glyph_literal(cursor) {
		Some(glyph)
	} else if let Some(identifier) = scan_identifier(cursor) {
		Some(identifier)
	} else {
		None
	}
}

fn scan_identifier(cursor: &mut Cursor) -> Option<Token> {
	let starting_predicate = |c: char| c.is_alphabetic() || c == '_';
	let suffix_predicate = |c: char| c.is_alphanumeric() || c == '_';

	if !cursor.match_char_predicate(&starting_predicate) {
		return None;
	}

	let mut lexeme = String::new();
	while cursor.current().is_some() && cursor.match_char_predicate(&suffix_predicate) {
		lexeme.push(cursor.next().unwrap());
	}

	if lexeme.len() == 1 && lexeme.starts_with('_') {
		Some(Token::new(TokenKind::UnderScore, cursor.len()))
	} else if let Some(keyword) = match_identifier_to_keyword(&lexeme) {
		Some(Token::new(keyword, cursor.len()))
	} else {
		Some(Token::new(TokenKind::Identifier, cursor.len()))
	}
}

fn scan_string_literal(cursor: &mut Cursor) -> Option<Token> {
	if !cursor.match_char('"') {
		return None;
	}

	cursor.next();
	cursor.next();

	while cursor.current().is_some() && cursor.match_char_predicate(&|c: char| c != '"') {
		cursor.next();
	}

	if cursor.match_char('"') {
		cursor.next();
		Some(Token::new(TokenKind::StringLiteral, cursor.len()))
	} else {
		Some(Token::new(TokenKind::TombStone, cursor.len()))
	}
}

fn scan_glyph_literal(cursor: &mut Cursor) -> Option<Token> {
	if !cursor.match_char('\'') {
		return None;
	}

	cursor.next();
	cursor.next();

	while cursor.current().is_some() && cursor.match_char_predicate(&|c: char| c != '\'') {
		cursor.next();
	}

	if cursor.match_char('\'') {
		cursor.next();
		Some(Token::new(TokenKind::GlyphLiteral, cursor.len()))
	} else {
		Some(Token::new(TokenKind::TombStone, cursor.len()))
	}
}

fn scan_numeric_literal(cursor: &mut Cursor) -> Option<Token> {
	if !cursor.match_char_predicate(&|c: char| c.is_digit(10)) {
		return None;
	}

	while cursor.current().is_some() && cursor.match_char_predicate(&|c: char| c.is_digit(10)) {
		cursor.next();
	}

	if cursor.match_char('.') {
		cursor.next();

		while cursor.current().is_some() && cursor.match_char_predicate(&|c: char| c.is_digit(10)) {
			cursor.next();
		}

		Some(Token::new(TokenKind::FloatLiteral, cursor.len()))
	} else {
		Some(Token::new(TokenKind::IntegerLiteral, cursor.len()))
	}
}


fn match_identifier_to_keyword(lexeme: &str) -> Option<TokenKind> {
	match lexeme.to_lowercase().as_str() {
		"fn" => Some(TokenKind::Fn),
		"return" => Some(TokenKind::Return),
		"if" => Some(TokenKind::If),
		"else" => Some(TokenKind::Else),
		"for" => Some(TokenKind::For),
		"in" => Some(TokenKind::In),
		"while" => Some(TokenKind::While),
		"let" => Some(TokenKind::Let),
		"true" => Some(TokenKind::True),
		"false" => Some(TokenKind::False),
		_ => None
	}
}
