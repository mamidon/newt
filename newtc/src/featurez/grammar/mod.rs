use self::expr::*;
use crate::featurez::parse::Parser;

mod expr {
    use crate::featurez::parse::Parser;
	use crate::featurez::parse::CompletedMarker;
    use crate::featurez::syntax::{SyntaxElement, SyntaxKind, SyntaxNode, SyntaxToken};
    use crate::featurez::{Token, TokenKind};
	use std::collections::HashMap;

	type BinaryOperatorHandler =  fn(&mut Parser, CompletedMarker) -> CompletedMarker;
	type UnaryOperatorHandler = fn(&mut Parser) -> CompletedMarker;
	type PrecedenceTable = HashMap<TokenKind, (usize, bool, bool)>;
	
	lazy_static! {
		static ref PRECEDENCE_TABLE: PrecedenceTable = [
			(TokenKind::IntegerLiteral, 0, false, true),
			(TokenKind::Star, 2, true, false),
			(TokenKind::Plus, 3, true, false),
			(TokenKind::Minus, 3, true, true),
			(TokenKind::Bang, 1, false, true),
		].iter()
			.map(|tuple| (tuple.0, (tuple.1, tuple.2, tuple.3)))
			.collect::<HashMap<_,_>>();
	}
	
    pub fn expr(p: &mut Parser) {
		let lhs = primary_expr(p);
		
		expr_core(p, lhs, 3);
    }
	
	// https://en.wikipedia.org/wiki/Operator-precedence_parser#Example_execution_of_the_algorithm
	// I've got the sort of right idea here -- but my notion of precedence is backwards 
	// additionally I only need an explicit function for parsing primary expressions, but I 
	// do need to encode the precedence & associativity of operators
	// also see http://craftinginterpreters.com/compiling-expressions.html#a-pratt-parser
	fn expr_core(p: &mut Parser, first_lhs: CompletedMarker, precedence: usize) -> CompletedMarker {
		let mut lookahead = p.current();
		let mut lhs = first_lhs;
		
		while lookahead.is_binary_operator() && lookahead_precedence(lookahead) <= precedence {
			let mut node = p.begin_node();
			p.precede_node(&mut lhs, &node);
			p.token_if(lookahead);
			let rhs = primary_expr(p);

			lookahead = p.current();

			if lookahead.is_binary_operator() && lookahead_precedence(lookahead) < precedence {
				expr_core(p, rhs, lookahead_precedence(lookahead));
				
				lookahead = p.current();
			}
			
			lhs = p.end_node(node, SyntaxKind::BinaryExpr);
		}
		
		lhs
	}
	
	fn lookahead_precedence(token: TokenKind) -> usize {
		PRECEDENCE_TABLE[&token].0
	}
	
	fn lookahead_binary(token: TokenKind) -> bool {
		PRECEDENCE_TABLE[&token].1
	}

	fn lookahead_unary(token: TokenKind) -> bool {
		PRECEDENCE_TABLE[&token].2
	}

	fn primary_expr(p: &mut Parser) -> CompletedMarker {
		let mut node = p.begin_node();
		
		let completed = match (p.nth(0), p.nth(1)) {
			(TokenKind::Minus, TokenKind::IntegerLiteral) => {
				p.token_if(TokenKind::Minus);
				p.token_if(TokenKind::IntegerLiteral);
				p.end_node(node, SyntaxKind::LiteralExpr)
			}, 
			(TokenKind::IntegerLiteral, _) => {
				p.token_if(TokenKind::IntegerLiteral);
				p.end_node(node, SyntaxKind::LiteralExpr)
			},
			(TokenKind::LeftParenthesis, _) => {
				p.expect_token_kind(TokenKind::LeftParenthesis, "Shouldn't happen");
				
				expr(p);
				
				p.expect_token_kind(TokenKind::RightParenthesis, "Expected ')' to close grouping");
				p.end_node(node, SyntaxKind::GroupingExpr)
			},
			_ => {
				p.expect_token_kind(TokenKind::IntegerLiteral, "Shouldn't happen");
				p.end_node(node, SyntaxKind::LiteralExpr)
			}
		};

		completed		
	}
}

pub fn root(p: &mut Parser) {
    expr(p);
}
