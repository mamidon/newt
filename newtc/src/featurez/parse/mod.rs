use crate::featurez::syntax::{
	BinaryExprNode, LiteralExprNode, SyntaxElement, SyntaxKind, SyntaxNode, SyntaxToken,
	SyntaxTree, TreeSink, TokenSource, TreeSink,
};
use crate::featurez::tokens::{tokenize, StrTokenSource, Token, TokenKind};
use std::fmt::Display;
use std::fmt::Error;
use std::fmt::Formatter;

mod parse_event;
pub use self::parse_event::ParseEvent;

mod marker;
use self::marker::Marker;

mod parser;
pub use self::parser::Parser;

pub fn parse(text: &str) -> SyntaxTree {
	use super::grammar::root;
	
    let tokens = tokenize(text);
    let source = StrTokenSource::new(tokens);
    let mut parser = Parser::new(text, source);

    root(&mut parser);
	
	let tree = SyntaxTree::from_parser(parser, text);
	
	tree
}

