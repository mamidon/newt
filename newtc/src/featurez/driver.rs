use crate::featurez::runtime::{VirtualMachineState, VirtualMachineInterpretingSession};
use crate::featurez::syntax::{AstNode, ExprNode, ExprVisitor, NewtValue, NewtStaticError, NewtRuntimeError, SyntaxNode};
use crate::featurez::syntax::{StmtNode, StmtVisitor, SyntaxElement, SyntaxTree};
use crate::featurez::tokens::{tokenize, StrTokenSource, Token, TokenKind};
use crate::featurez::parse::Parser;

use std::fmt::Debug;
use std::fmt::Display;
use std::fmt::Error;
use std::fmt::Formatter;
use std::collections::HashMap;
use crate::featurez::grammar::root_stmt;


#[derive(Copy, Clone)]
pub enum InterpretingSessionKind {
	Stmt,
	Expr,
}

pub enum NewtError {
	Static(NewtStaticError),
	Runtime(NewtRuntimeError)
}

pub struct InterpretingSession<'sess> {
	kind: InterpretingSessionKind,
	source: &'sess str,
	tree: SyntaxTree,
}

impl<'sess> InterpretingSession<'sess> {
	pub fn new(kind: InterpretingSessionKind, source: &'sess str) -> InterpretingSession<'sess> {
		let mut tree = InterpretingSession::syntax_tree_from_source(kind, source);

		InterpretingSession {
			kind,
			source,
			tree,
		}
	}

	pub fn interpret(&self, vm: &mut VirtualMachineState) -> Option<NewtValue> {
		let mut session = VirtualMachineInterpretingSession::new(self.syntax_tree(), vm);
		session.interpret()
	}

	pub fn syntax_tree(&self) -> &SyntaxTree {
		&self.tree
	}

	fn syntax_tree_from_source(kind: InterpretingSessionKind, source: &'sess str) -> SyntaxTree {
		use super::grammar::{root_expr, root_stmt};

		let tokens = tokenize(source);
		let token_source = StrTokenSource::new(tokens);
		let mut parser = Parser::new(token_source);

		let completed_parsing = match kind {
			InterpretingSessionKind::Stmt => root_stmt(parser),
			InterpretingSessionKind::Expr => root_expr(parser),
		};

		SyntaxTree::from_parser(&completed_parsing, source)
	}
}


impl From<&str> for SyntaxTree {
	fn from(source: &str) -> Self {
		let tokens = tokenize(source);
		let token_source = StrTokenSource::new(tokens);
		let mut p = Parser::new(token_source);

		SyntaxTree::from_parser(&root_stmt(p), source)
	}
}


