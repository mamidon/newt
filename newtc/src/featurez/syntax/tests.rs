#![cfg(test)]

use super::*;

use insta::assert_snapshot_matches;

use crate::featurez::grammar::{root_expr, root_stmt};
use crate::featurez::parse::Parser;
use crate::featurez::tokens::tokenize;
use crate::featurez::tokens::StrTokenSource;
use crate::featurez::tokens::TokenKind;

macro_rules! syntax_tree_expr_tests {
	($($name:ident: $test_source:expr,)*) => {
	$(
		#[test]
		fn $name() {
			let text: &str = $test_source;
			let tokens = tokenize(text);
			let parser = Parser::new(StrTokenSource::new(tokens.clone()));
			let completed_parsing = root_expr(parser);
			let tree = SyntaxTree::from_parser(&completed_parsing, text);

			let approval_document = format!("====text====\n============\n{}\n\
				====tokens====\n============\n{:#?}\n\
				====events====\n============\n{:#?}\n\
				====tree====\n============\n{:#?}", text, tokens, completed_parsing.events, tree);
	assert_snapshot_matches!(stringify!($name), approval_document);
		}
	)*
	}
}

macro_rules! syntax_tree_stmt_tests {
	($($name:ident: $test_source:expr,)*) => {
	$(
		#[test]
		fn $name() {
			let text: &str = $test_source;
			let tokens = tokenize(text);
			let mut parser = Parser::new(StrTokenSource::new(tokens.clone()));
			let completed_parsing = root_stmt(parser);
			let tree = SyntaxTree::from_parser(&completed_parsing, text);

			let approval_document = format!("====text====\n============\n{}\n\
				====tokens====\n============\n{:#?}\n\
				====events====\n============\n{:#?}\n\
				====tree====\n============\n{:#?}", text, tokens, completed_parsing.events, tree);
	assert_snapshot_matches!(stringify!($name), approval_document);
		}
	)*
	}
}

// associativity & precedence
syntax_tree_expr_tests! {
    left_associativity_is_deeply_nested: "1+2+3",
    higher_precedence_is_evaluated_first: "1+2*3",
    higher_precedence_is_noop_when_first: "1*2+3",
    unary_operators_are_properly_grouped: "-1*2+-3",
    nested_unary_operators: "1*--2.12",
    grouping_is_highest_precedence: "(1+2)*3",
    expr_starting_whitespace_is_fine: r#"
    1*2"#,
}

syntax_tree_stmt_tests! {
    stmt_starting_whitespace_is_fine: r#"
	let x = 1;
	let y = 2;
	let c = x * 2 + y;
	"#,
}

// operators
