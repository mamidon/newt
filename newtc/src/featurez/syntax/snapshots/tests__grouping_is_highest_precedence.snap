---
created: "2019-07-29T15:38:38.653649Z"
creator: insta@0.8.1
source: newtc/src/featurez/syntax/tests.rs
expression: approval_document
---
(1+2)*3
===>
┗ BinaryExpr
  ┠ GroupingExpr
  ┃ ┠ LeftParenthesis "("
  ┃ ┠ BinaryExpr
  ┃ ┃ ┠ LiteralExpr
  ┃ ┃ ┃ ┗ IntegerLiteral "1"
  ┃ ┃ ┠ Plus "+"
  ┃ ┃ ┗ LiteralExpr
  ┃ ┃   ┗ IntegerLiteral "2"
  ┃ ┗ RightParenthesis ")"
  ┠ Star "*"
  ┗ LiteralExpr
    ┗ IntegerLiteral "3"

